File: pjecalc-fonte/negocio/br/jus/trt8/pjecalc/negocio/dominio/calculo/juros/ApuracaoDeJuros.java
/*
 * Decompiled with CFR 0.152.
 * 
 * Could not load the following classes:
 *  javax.persistence.Column
 *  javax.persistence.Entity
 *  javax.persistence.FetchType
 *  javax.persistence.GeneratedValue
 *  javax.persistence.GenerationType
 *  javax.persistence.Id
 *  javax.persistence.JoinColumn
 *  javax.persistence.ManyToOne
 *  javax.persistence.SequenceGenerator
 *  javax.persistence.Table
 *  javax.persistence.Temporal
 *  javax.persistence.TemporalType
 *  javax.persistence.Version
 *  org.jboss.seam.ScopeType
 *  org.jboss.seam.annotations.Name
 *  org.jboss.seam.annotations.Scope
 */
package br.jus.trt8.pjecalc.negocio.dominio.calculo.juros;

import br.jus.trt8.pjecalc.base.comum.EntidadeBase;
import br.jus.trt8.pjecalc.base.comum.Utils;
import br.jus.trt8.pjecalc.negocio.dominio.calculo.Calculo;
import br.jus.trt8.pjecalc.negocio.dominio.calculo.juros.RepositorioDeApuracaoDeJuros;
import java.math.BigDecimal;
import java.util.Date;
import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.FetchType;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.JoinColumn;
import javax.persistence.ManyToOne;
import javax.persistence.SequenceGenerator;
import javax.persistence.Table;
import javax.persistence.Temporal;
import javax.persistence.TemporalType;
import javax.persistence.Version;
import org.jboss.seam.ScopeType;
import org.jboss.seam.annotations.Name;
import org.jboss.seam.annotations.Scope;

@Entity
@Table(name="TBAPURACAOJUROSCALCULO")
@SequenceGenerator(name="SQAPURACAOJUROSCALCULO", sequenceName="SQAPURACAOJUROSCALCULO", allocationSize=1)
@Scope(value=ScopeType.SESSION)
@Name(value="apuracaoDeJuros")
public class ApuracaoDeJuros
extends EntidadeBase {
    private static final long serialVersionUID = 6872629805146414382L;
    @Id
    @GeneratedValue(strategy=GenerationType.SEQUENCE, generator="SQAPURACAOJUROSCALCULO")
    @Column(name="IIDOCORRENCIAAPURACAO")
    private final Long id = null;
    @Version
    @Column(name="IIDVERSAO")
    private Long versao = 0L;
    @ManyToOne(fetch=FetchType.LAZY)
    @JoinColumn(name="IIDCALCULO")
    private Calculo calculo;
    @Column(name="DDTOCORRENCIA", nullable=false)
    @Temporal(value=TemporalType.DATE)
    private Date competencia;
    @Column(name="DDTINICIAL", nullable=false)
    @Temporal(value=TemporalType.DATE)
    private Date dataInicial;
    @Column(name="RVLCORRIGIDOVERBA", precision=19, scale=2)
    private BigDecimal valorCorrigido = BigDecimal.ZERO;
    @Column(name="RVLVERBACONTRSOCIAL", precision=19, scale=2)
    private BigDecimal valorVerbaParaContribuicaoSocial = BigDecimal.ZERO;
    @Column(name="RVLVERBACONTRSOCIAL13", precision=19, scale=2)
    private BigDecimal valorVerbaParaContribuicaoSocialDecimoTerceiro = BigDecimal.ZERO;
    @Column(name="RVLVERBAPREVPRIVADA", precision=19, scale=2)
    private BigDecimal valorVerbaParaPrevidenciaPrivada = BigDecimal.ZERO;
    @Column(name="RVLCORRIGIDOVERBAIRPFDECIMO", precision=19, scale=2)
    private BigDecimal valorCorrigidoParaIrpfDecimoTerceiro = BigDecimal.ZERO;
    @Column(name="RVLCORRIGIDOVERBAIRPFFERIAS", precision=19, scale=2)
    private BigDecimal valorCorrigidoParaIrpfFerias = BigDecimal.ZERO;
    @Column(name="RVLCORRIGIDOVERBAIRPFDEMAIS", precision=19, scale=2)
    private BigDecimal valorCorrigidoParaIrpfDemaisVerbas = BigDecimal.ZERO;
    @Column(name="RVLCORRIGIDOCONTRIBUICAOSOCIAL", precision=19, scale=2)
    private BigDecimal contribuicaoSocialNormal = BigDecimal.ZERO;
    @Column(name="RVLCORRIGIDOCONTRSOCIAL13", precision=19, scale=2)
    private BigDecimal contribuicaoSocialDecimoTerceiro = BigDecimal.ZERO;
    @Column(name="RVLCORRIGIDOPREVIDENCIAPRIVADA", precision=19, scale=2)
    private BigDecimal previdenciaPrivada = BigDecimal.ZERO;
    @Column(name="RVLTAXAJUROS", precision=38, scale=25)
    private BigDecimal taxaDeJuros;

    public ApuracaoDeJuros() {
        super(RepositorioDeApuracaoDeJuros.class);
    }

    public ApuracaoDeJuros(Calculo calculo) {
        this();
        this.calculo = calculo;
    }

    @Override
    public Object obterChavePrimaria() {
        return this.getId();
    }

    public Long getId() {
        return this.id;
    }

    public Date getCompetencia() {
        return this.competencia;
    }

    @Override
    public Long getVersao() {
        return this.versao;
    }

    @Override
    public void setVersao(Long versao) {
        this.versao = versao;
    }

    public void setCompetencia(Date competencia) {
        this.competencia = competencia;
    }

    public Date getDataInicial() {
        return this.dataInicial;
    }

    public void setDataInicial(Date dataInicial) {
        this.dataInicial = dataInicial;
    }

    public Calculo getCalculo() {
        return this.calculo;
    }

    public void setCalculo(Calculo calculo) {
        this.calculo = calculo;
    }

    public BigDecimal getValorCorrigido() {
        return this.valorCorrigido;
    }

    public void setValorCorrigido(BigDecimal valorCorrigido) {
        this.valorCorrigido = valorCorrigido;
    }

    public BigDecimal getValorVerbaParaContribuicaoSocial() {
        return this.valorVerbaParaContribuicaoSocial;
    }

    public void setValorVerbaParaContribuicaoSocial(BigDecimal valorVerbaParaContribuicaoSocial) {
        this.valorVerbaParaContribuicaoSocial = valorVerbaParaContribuicaoSocial;
    }

    public BigDecimal getValorVerbaParaContribuicaoSocialDecimoTerceiro() {
        return this.valorVerbaParaContribuicaoSocialDecimoTerceiro;
    }

    public void setValorVerbaParaContribuicaoSocialDecimoTerceiro(BigDecimal valorVerbaParaContribuicaoSocialDecimoTerceiro) {
        this.valorVerbaParaContribuicaoSocialDecimoTerceiro = valorVerbaParaContribuicaoSocialDecimoTerceiro;
    }

    public BigDecimal getValorVerbaParaPrevidenciaPrivada() {
        return this.valorVerbaParaPrevidenciaPrivada;
    }

    public void setValorVerbaParaPrevidenciaPrivada(BigDecimal valorVerbaParaPrevidenciaPrivada) {
        this.valorVerbaParaPrevidenciaPrivada = valorVerbaParaPrevidenciaPrivada;
    }

    public BigDecimal getValorCorrigidoParaIrpfDecimoTerceiro() {
        return this.valorCorrigidoParaIrpfDecimoTerceiro;
    }

    public void setValorCorrigidoParaIrpfDecimoTerceiro(BigDecimal valorCorrigidoParaIrpfDecimoTerceiro) {
        this.valorCorrigidoParaIrpfDecimoTerceiro = valorCorrigidoParaIrpfDecimoTerceiro;
    }

    public BigDecimal getValorCorrigidoParaIrpfFerias() {
        return this.valorCorrigidoParaIrpfFerias;
    }

    public void setValorCorrigidoParaIrpfFerias(BigDecimal valorCorrigidoParaIrpfFerias) {
        this.valorCorrigidoParaIrpfFerias = valorCorrigidoParaIrpfFerias;
    }

    public BigDecimal getValorCorrigidoParaIrpfDemaisVerbas() {
        return this.valorCorrigidoParaIrpfDemaisVerbas;
    }

    public void setValorCorrigidoParaIrpfDemaisVerbas(BigDecimal valorCorrigidoParaIrpfDemaisVerbas) {
        this.valorCorrigidoParaIrpfDemaisVerbas = valorCorrigidoParaIrpfDemaisVerbas;
    }

    public BigDecimal getContribuicaoSocialNormal() {
        return this.contribuicaoSocialNormal;
    }

    public void setContribuicaoSocialNormal(BigDecimal contribuicaoSocialNormal) {
        this.contribuicaoSocialNormal = contribuicaoSocialNormal;
    }

    public BigDecimal getContribuicaoSocialDecimoTerceiro() {
        return this.contribuicaoSocialDecimoTerceiro;
    }

    public void setContribuicaoSocialDecimoTerceiro(BigDecimal contribuicaoSocialDecimoTerceiro) {
        this.contribuicaoSocialDecimoTerceiro = contribuicaoSocialDecimoTerceiro;
    }

    public BigDecimal getContribuicaoSocial() {
        return Utils.somar(this.contribuicaoSocialNormal, this.contribuicaoSocialDecimoTerceiro);
    }

    public BigDecimal getPrevidenciaPrivada() {
        return this.previdenciaPrivada;
    }

    public void setPrevidenciaPrivada(BigDecimal previdenciaPrivada) {
        this.previdenciaPrivada = previdenciaPrivada;
    }

    public BigDecimal getTaxaDeJuros() {
        return this.taxaDeJuros;
    }

    public void setTaxaDeJuros(BigDecimal taxaDeJuros) {
        this.taxaDeJuros = taxaDeJuros;
    }

    public BigDecimal getCapital() {
        if (Utils.naoNulo(this.valorCorrigido)) {
            BigDecimal capital = this.valorCorrigido;
            if (Utils.naoNulo(this.getContribuicaoSocial())) {
                capital = capital.subtract(this.getContribuicaoSocial(), Utils.CONTEXTO_MATEMATICO);
            }
            if (Utils.naoNulo(this.previdenciaPrivada)) {
                capital = capital.subtract(this.previdenciaPrivada, Utils.CONTEXTO_MATEMATICO);
            }
            return capital;
        }
        return null;
    }

    public BigDecimal getJuros() {
        BigDecimal capital = this.getCapital();
        if (Utils.naoNulos(capital, this.taxaDeJuros)) {
            return capital.multiply(Utils.obterPercentualPara(this.taxaDeJuros), Utils.CONTEXTO_MATEMATICO);
        }
        return null;
    }

    public BigDecimal getTotal() {
        if (Utils.naoNulos(this.getCapital(), this.getJuros())) {
            return this.getCapital().add(this.getJuros(), Utils.CONTEXTO_MATEMATICO);
        }
        return null;
    }

    public BigDecimal getJurosParaIrpfDecimoTerceiro() {
        if (BigDecimal.ZERO.compareTo(this.getValorCorrigido()) == 0) {
            return BigDecimal.ZERO;
        }
        BigDecimal jurosIrpfDecimoTerceiro = this.getJuros();
        jurosIrpfDecimoTerceiro = jurosIrpfDecimoTerceiro.multiply(this.getValorCorrigidoParaIrpfDecimoTerceiro(), Utils.CONTEXTO_MATEMATICO);
        jurosIrpfDecimoTerceiro = jurosIrpfDecimoTerceiro.divide(this.getValorCorrigido(), Utils.CONTEXTO_MATEMATICO);
        return jurosIrpfDecimoTerceiro;
    }

    public BigDecimal getJurosParaIrpfFerias() {
        if (BigDecimal.ZERO.compareTo(this.getValorCorrigido()) == 0) {
            return BigDecimal.ZERO;
        }
        BigDecimal jurosIrpfFerias = this.getJuros();
        jurosIrpfFerias = jurosIrpfFerias.multiply(this.getValorCorrigidoParaIrpfFerias(), Utils.CONTEXTO_MATEMATICO);
        jurosIrpfFerias = jurosIrpfFerias.divide(this.getValorCorrigido(), Utils.CONTEXTO_MATEMATICO);
        return jurosIrpfFerias;
    }

    public BigDecimal getJurosParaIrpfDemaisVerbas() {
        if (BigDecimal.ZERO.compareTo(this.getValorCorrigido()) == 0) {
            return BigDecimal.ZERO;
        }
        BigDecimal jurosIrpfDemaisVerbas = this.getJuros();
        jurosIrpfDemaisVerbas = jurosIrpfDemaisVerbas.multiply(this.getValorCorrigidoParaIrpfDemaisVerbas(), Utils.CONTEXTO_MATEMATICO);
        jurosIrpfDemaisVerbas = jurosIrpfDemaisVerbas.divide(this.getValorCorrigido(), Utils.CONTEXTO_MATEMATICO);
        return jurosIrpfDemaisVerbas;
    }

    public void remover() {
        ApuracaoDeJuros.getRepositorio(RepositorioDeApuracaoDeJuros.class).remover(this, false);
    }
}

File: src/lib/pjecalc/core/dominio/calculo/juros/apuracao-de-juros.ts
/**
 * PJe-Calc v2.15.1 — ApuracaoDeJuros
 * Porte 1:1 de: br.jus.trt8.pjecalc.negocio.dominio.calculo.juros.ApuracaoDeJuros
 *
 * Ref Java: pjecalc-fonte/negocio/br/jus/trt8/pjecalc/negocio/dominio/calculo/juros/ApuracaoDeJuros.java
 *
 * Linha-mensal da apuração de juros do cálculo. Para cada competência do
 * `Calculo`, armazena:
 *   - valorCorrigido (total do mês, já com correção monetária)
 *   - valores base do INSS (CS normal + CS 13º) e da Prev. Privada
 *   - valores corrigidos separados para IRPF (13º, férias, demais)
 *   - contribuição social e previdência privada mensais (parcelas já deduzidas)
 *   - taxa de juros da competência (% mensal)
 *
 * Derivados:
 *   getCapital()  — valorCorrigido − contribuiçãoSocial − previdenciaPrivada
 *   getJuros()    — capital × (taxa / 100)
 *   getTotal()    — capital + juros
 *   getJurosParaIrpf{DecimoTerceiro,Ferias,DemaisVerbas}() — juros rateados
 *                    proporcionalmente ao componente de IRPF.
 */
import Decimal from 'decimal.js';
import type { Calculo } from '../calculo';

const ZERO = new Decimal(0);

function somar(a: Decimal | null | undefined, b: Decimal | null | undefined): Decimal {
  const va = a ?? ZERO;
  const vb = b ?? ZERO;
  return va.plus(vb);
}

/** Utils.obterPercentualPara (Java) — transforma taxa em % para fator decimal. */
function obterPercentualPara(taxa: Decimal): Decimal {
  return taxa.div(100);
}

export class ApuracaoDeJuros {
  private id: number | null = null;
  private versao: number = 0;
  private calculo: Calculo | null;
  private competencia: Date | null = null;
  private dataInicial: Date | null = null;
  private valorCorrigido: Decimal = ZERO;
  private valorVerbaParaContribuicaoSocial: Decimal = ZERO;
  private valorVerbaParaContribuicaoSocialDecimoTerceiro: Decimal = ZERO;
  private valorVerbaParaPrevidenciaPrivada: Decimal = ZERO;
  private valorCorrigidoParaIrpfDecimoTerceiro: Decimal = ZERO;
  private valorCorrigidoParaIrpfFerias: Decimal = ZERO;
  private valorCorrigidoParaIrpfDemaisVerbas: Decimal = ZERO;
  private contribuicaoSocialNormal: Decimal = ZERO;
  private contribuicaoSocialDecimoTerceiro: Decimal = ZERO;
  private previdenciaPrivada: Decimal = ZERO;
  private taxaDeJuros: Decimal | null = null;

  constructor(calculo: Calculo | null = null) {
    this.calculo = calculo;
  }

  getId(): number | null { return this.id; }
  setId(id: number): void { this.id = id; }
  getVersao(): number { return this.versao; }
  setVersao(v: number): void { this.versao = v; }

  getCalculo(): Calculo | null { return this.calculo; }
  setCalculo(c: Calculo | null): void { this.calculo = c; }

  getCompetencia(): Date | null { return this.competencia; }
  setCompetencia(c: Date): void { this.competencia = c; }

  getDataInicial(): Date | null { return this.dataInicial; }
  setDataInicial(d: Date): void { this.dataInicial = d; }

  getValorCorrigido(): Decimal { return this.valorCorrigido; }
  setValorCorrigido(v: Decimal): void { this.valorCorrigido = v; }

  getValorVerbaParaContribuicaoSocial(): Decimal { return this.valorVerbaParaContribuicaoSocial; }
  setValorVerbaParaContribuicaoSocial(v: Decimal): void { this.valorVerbaParaContribuicaoSocial = v; }

  getValorVerbaParaContribuicaoSocialDecimoTerceiro(): Decimal {
    return this.valorVerbaParaContribuicaoSocialDecimoTerceiro;
  }
  setValorVerbaParaContribuicaoSocialDecimoTerceiro(v: Decimal): void {
    this.valorVerbaParaContribuicaoSocialDecimoTerceiro = v;
  }

  getValorVerbaParaPrevidenciaPrivada(): Decimal { return this.valorVerbaParaPrevidenciaPrivada; }
  setValorVerbaParaPrevidenciaPrivada(v: Decimal): void { this.valorVerbaParaPrevidenciaPrivada = v; }

  getValorCorrigidoParaIrpfDecimoTerceiro(): Decimal { return this.valorCorrigidoParaIrpfDecimoTerceiro; }
  setValorCorrigidoParaIrpfDecimoTerceiro(v: Decimal): void { this.valorCorrigidoParaIrpfDecimoTerceiro = v; }

  getValorCorrigidoParaIrpfFerias(): Decimal { return this.valorCorrigidoParaIrpfFerias; }
  setValorCorrigidoParaIrpfFerias(v: Decimal): void { this.valorCorrigidoParaIrpfFerias = v; }

  getValorCorrigidoParaIrpfDemaisVerbas(): Decimal { return this.valorCorrigidoParaIrpfDemaisVerbas; }
  setValorCorrigidoParaIrpfDemaisVerbas(v: Decimal): void { this.valorCorrigidoParaIrpfDemaisVerbas = v; }

  getContribuicaoSocialNormal(): Decimal { return this.contribuicaoSocialNormal; }
  setContribuicaoSocialNormal(v: Decimal): void { this.contribuicaoSocialNormal = v; }

  getContribuicaoSocialDecimoTerceiro(): Decimal { return this.contribuicaoSocialDecimoTerceiro; }
  setContribuicaoSocialDecimoTerceiro(v: Decimal): void { this.contribuicaoSocialDecimoTerceiro = v; }

  /** getContribuicaoSocial (Java linha 218) — soma normal + 13º */
  getContribuicaoSocial(): Decimal {
    return somar(this.contribuicaoSocialNormal, this.contribuicaoSocialDecimoTerceiro);
  }

  getPrevidenciaPrivada(): Decimal { return this.previdenciaPrivada; }
  setPrevidenciaPrivada(v: Decimal): void { this.previdenciaPrivada = v; }

  getTaxaDeJuros(): Decimal | null { return this.taxaDeJuros; }
  setTaxaDeJuros(t: Decimal | null): void { this.taxaDeJuros = t; }

  /**
   * getCapital (Java linha 238)
   * Capital = valorCorrigido − (CS normal + CS 13º) − previdenciaPrivada
   * Retorna null se valorCorrigido for null (paridade com Java).
   */
  getCapital(): Decimal | null {
    if (this.valorCorrigido === null || this.valorCorrigido === undefined) return null;
    let capital = this.valorCorrigido;
    const cs = this.getContribuicaoSocial();
    if (cs !== null && cs !== undefined) {
      capital = capital.minus(cs);
    }
    if (this.previdenciaPrivada !== null && this.previdenciaPrivada !== undefined) {
      capital = capital.minus(this.previdenciaPrivada);
    }
    return capital;
  }

  /** getJuros (Java linha 252) — capital × (taxaDeJuros / 100). Null se capital ou taxa nulos. */
  getJuros(): Decimal | null {
    const capital = this.getCapital();
    if (capital === null || this.taxaDeJuros === null) return null;
    return capital.times(obterPercentualPara(this.taxaDeJuros));
  }

  /** getTotal (Java linha 260) — capital + juros */
  getTotal(): Decimal | null {
    const capital = this.getCapital();
    const juros = this.getJuros();
    if (capital === null || juros === null) return null;
    return capital.plus(juros);
  }

  /**
   * getJurosParaIrpfDecimoTerceiro (Java linha 267) — juros rateados
   * proporcionalmente ao componente de 13º dentro de valorCorrigido.
   */
  getJurosParaIrpfDecimoTerceiro(): Decimal {
    if (this.valorCorrigido.isZero()) return ZERO;
    const juros = this.getJuros() ?? ZERO;
    return juros.times(this.valorCorrigidoParaIrpfDecimoTerceiro).div(this.valorCorrigido);
  }

  /** getJurosParaIrpfFerias (Java linha 277) */
  getJurosParaIrpfFerias(): Decimal {
    if (this.valorCorrigido.isZero()) return ZERO;
    const juros = this.getJuros() ?? ZERO;
    return juros.times(this.valorCorrigidoParaIrpfFerias).div(this.valorCorrigido);
  }

  /** getJurosParaIrpfDemaisVerbas (Java linha 287) */
  getJurosParaIrpfDemaisVerbas(): Decimal {
    if (this.valorCorrigido.isZero()) return ZERO;
    const juros = this.getJuros() ?? ZERO;
    return juros.times(this.valorCorrigidoParaIrpfDemaisVerbas).div(this.valorCorrigido);
  }
}
